library/x509_create.c

```diff
@@ -123,53 +130,199 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name(const char *name,
+        mbedtls_free(oid->p);
+        mbedtls_free(oid);
+        return NULL;
+    }
+
+    for (cur = x509_attrs; cur->oid != NULL; cur++) {
+        if (strlen(cur->oid) == oid->len &&
+            strncmp(cur->oid, (const char *) oid->p, oid->len) == 0) {
+            break;
+        }
+    }
+
+    mbedtls_free(oid->p);
+    mbedtls_free(oid);
+    if (cur->oid == NULL) {
+        return NULL;
+    }
+
+    return cur;
+}
+
+static int hex_to_int(char c)
+{
+    return ('0' <= c && c <= '9') ? (c - '0') :
+           ('a' <= c && c <= 'f') ? (c - 'a' + 10) :
+           ('A' <= c && c <= 'F') ? (c - 'A' + 10) : -1;
+}
+
+static int hexpair_to_int(char c1, char c2)
+{
+    int n1 = hex_to_int(c1);
+    int n2 = hex_to_int(c2);
+    if (n1 != -1 && n2 != -1) {
+        return (n1 << 4) | n2;
+    } else {
+        return -1;
+    }
+}
+
+static int parse_attribute_value_string(const char *s,
+                                        int len,
+                                        unsigned char *data,
+                                        size_t *data_len)
+{
+    const char *c = s;
+    const char *end = c + len;
+    int hexpair = 0;
+    unsigned char *d = data;
+    int n;
+    while (c < end) {
+        if (*c == '\\') {
+            c++;
+
+            /* Check for valid escaped characters in RFC 4514 in Section 3*/
+            if (c + 1 < end && (n = hexpair_to_int(*c, *(c+1))) != -1) {
+                hexpair = 1;
+                *(d++) = n;
+                c++;
+            } else if (c == end || !strchr(" ,=+<>#;\"\\+", *c)) {
+                return MBEDTLS_ERR_X509_INVALID_NAME;
+            }
+        }
+        if (!hexpair) {
+            *(d++) = *c;
+        }
+        if (d - data == MBEDTLS_X509_MAX_DN_NAME_SIZE) {
+            return MBEDTLS_ERR_X509_INVALID_NAME;
+        }
+
+        hexpair = 0;
+        c++;
+    }
+    *data_len = d - data;
+    return 0;
+}
+
+#if defined(MBEDTLS_ASN1_PARSE_C)
+static int parse_attribute_value_ber_encoded(const char *s,
+                                             int len,
+                                             unsigned char *data,
+                                             size_t *data_len,
+                                             int *tag)
+{
+    const char *c = s;
+    const char *end = c + len;
+    unsigned char asn1_der_buf[256];
+    unsigned char *asn1_der_end;
+    unsigned char *p;
+    unsigned char *d;
+    int n;
+    /* Converting from hexstring to raw binary so we can use asn1parse.c*/
+    if ((len < 5) || (*c != '#')) {
+        return MBEDTLS_ERR_X509_INVALID_NAME;
+    }
+    c++;
+    if ((*tag = hexpair_to_int(*c, *(c+1))) == -1) {
+        return MBEDTLS_ERR_X509_INVALID_NAME;
+    }
+    c += 2;
+    p = asn1_der_buf;
+    for (p = asn1_der_buf; c < end; c += 2) {
+        if ((c + 1 >= end) || (n = hexpair_to_int(*c, *(c+1))) == -1) {
+            return MBEDTLS_ERR_X509_INVALID_NAME;
+        }
```

Comment:
For now I think we should explicitly reject escaped `'\0'` characters. Allowing the user to create certificates with strings that contain `'\0'` in the middle seems like a bad idea. UTF-8 work can address unacceptable characters more thoroughly, but it's best to have some safeguard.
```suggestion
        }
        if (n == '\0') {
            return MBEDTLS_ERR_X509_INVALID_NAME;
        }
```

Reply:
This function doesn't just parse for strings but any ber encoded data, I know some certificates in the testing framework have encoded bitstrings with null bytes.

Reply:
You're right, I meant to raise this [above](https://github.com/Mbed-TLS/mbedtls/pull/8025/files/d265d14c4b56959333d0e36e0b257a788fb22313#r1290375707), not here.